File: backend/core/src/modules/document-comments/document-comments.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { Document, DocumentComment } from '../../entities';

@Injectable()
export class DocumentCommentsService {
  constructor(
    @InjectRepository(DocumentComment)
    private readonly commentsRepo: Repository<DocumentComment>,
    @InjectRepository(Document)
    private readonly documentsRepo: Repository<Document>,
    private readonly projectAccessService: ProjectAccessService,
  ) {}

  private async assertDocumentAccess(userId: string, documentId: string): Promise<Document> {
    const doc = await this.documentsRepo.findOne({ where: { id: documentId } });
    if (!doc) {
      throw new BadRequestException('문서를 찾을 수 없습니다.');
    }
    await this.projectAccessService.assertProjectAccess(userId, doc.projectId);
    return doc;
  }

  async findAllForUser(userId: string, documentId: string) {
    await this.assertDocumentAccess(userId, documentId);
    return this.commentsRepo.find({
      where: { documentId } as any,
      order: { createdAt: 'ASC' } as any,
    });
  }

  async findOneForUser(userId: string, id: string) {
    const comment = await this.commentsRepo.findOne({ where: { id } });
    if (!comment) return null;
    await this.assertDocumentAccess(userId, comment.documentId);
    return comment;
  }

  async createForUser(userId: string, dto: Partial<DocumentComment>) {
    if (!dto.documentId) {
      throw new BadRequestException('documentId가 필요합니다.');
    }
    await this.assertDocumentAccess(userId, dto.documentId);
    const comment = this.commentsRepo.create({ ...dto, userId: dto.userId ?? userId } as any);
    return this.commentsRepo.save(comment as any);
  }

  async updateForUser(userId: string, id: string, dto: Partial<DocumentComment>) {
    const comment = await this.commentsRepo.findOne({ where: { id } });
    if (!comment) return null;
    await this.assertDocumentAccess(userId, comment.documentId);
    await this.commentsRepo.update({ id } as any, dto as any);
    return this.findOneForUser(userId, id);
  }

  async removeForUser(userId: string, id: string) {
    const comment = await this.commentsRepo.findOne({ where: { id } });
    if (!comment) return;
    await this.assertDocumentAccess(userId, comment.documentId);
    await this.commentsRepo.delete({ id } as any);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/DocumentCommentsService.java
package store.pjcloud.cowrite.core.service;

import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.entity.DocumentComment;
import store.pjcloud.cowrite.core.repository.DocumentCommentRepository;
import store.pjcloud.cowrite.core.repository.DocumentRepository;

@Service
public class DocumentCommentsService {
    private final DocumentCommentRepository commentsRepo;
    private final DocumentRepository documentsRepo;
    private final ProjectAccessService projectAccessService;

    public DocumentCommentsService(DocumentCommentRepository commentsRepo,
                                   DocumentRepository documentsRepo,
                                   ProjectAccessService projectAccessService) {
        this.commentsRepo = commentsRepo;
        this.documentsRepo = documentsRepo;
        this.projectAccessService = projectAccessService;
    }

    private Document assertDocumentAccess(UUID userId, UUID documentId) {
        Document doc = documentsRepo.findById(documentId).orElse(null);
        if (doc == null) {
            throw new IllegalArgumentException("문서를 찾을 수 없습니다.");
        }
        projectAccessService.assertProjectAccess(userId, doc.getProjectId());
        return doc;
    }

    public List<DocumentComment> findAllForUser(UUID userId, UUID documentId) {
        assertDocumentAccess(userId, documentId);
        return commentsRepo.findByDocumentIdOrderByCreatedAtAsc(documentId);
    }

    public DocumentComment findOneForUser(UUID userId, UUID id) {
        DocumentComment comment = commentsRepo.findById(id).orElse(null);
        if (comment == null) return null;
        assertDocumentAccess(userId, comment.getDocumentId());
        return comment;
    }

    @Transactional
    public DocumentComment createForUser(UUID userId, DocumentComment dto) {
        if (dto.getDocumentId() == null) {
            throw new IllegalArgumentException("documentId가 필요합니다.");
        }
        assertDocumentAccess(userId, dto.getDocumentId());
        if (dto.getUserId() == null) dto.setUserId(userId);
        return commentsRepo.save(dto);
    }

    @Transactional
    public DocumentComment updateForUser(UUID userId, UUID id, DocumentComment dto) {
        DocumentComment existing = commentsRepo.findById(id).orElse(null);
        if (existing == null) return null;
        assertDocumentAccess(userId, existing.getDocumentId());
        if (dto.getContent() != null) existing.setContent(dto.getContent());
        if (dto.getPosition() != null) existing.setPosition(dto.getPosition());
        return commentsRepo.save(existing);
    }

    @Transactional
    public void removeForUser(UUID userId, UUID id) {
        DocumentComment existing = commentsRepo.findById(id).orElse(null);
        if (existing == null) return;
        assertDocumentAccess(userId, existing.getDocumentId());
        commentsRepo.deleteById(id);
    }
}
